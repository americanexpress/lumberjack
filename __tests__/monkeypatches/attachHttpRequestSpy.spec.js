/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import http from 'http';

import onFinished from 'on-finished';

import attachHttpRequestSpy from '../../src/monkeypatches/attachHttpRequestSpy';

import attachSpy from '../../src/monkeypatches/attachSpy';

jest.mock('http', () => ({ request: jest.fn() }));
jest.mock('on-finished');
jest.mock('../../src/monkeypatches/attachSpy');

describe('attachHttpRequestSpy', () => {
  beforeEach(() => {
    attachSpy.mockClear();
  });

  it('throws if requestSpy is not a function', () => {
    expect(attachHttpRequestSpy).toThrowErrorMatchingSnapshot();
  });

  it('does not throw if socketCloseSpy is not provided', () => {
    expect(() => attachHttpRequestSpy(() => {})).not.toThrow();
  });

  it('throws if socketCloseSpy is provided but is not a function', () => {
    expect(() => attachHttpRequestSpy(() => {}, 'apples')).toThrowErrorMatchingSnapshot();
  });

  it('attaches the spy', () => {
    attachHttpRequestSpy(jest.fn());
    expect(attachSpy).toHaveBeenCalledTimes(1);
    expect(attachSpy.mock.calls[0][0]).toBe(http);
    expect(attachSpy.mock.calls[0][1]).toBe('request');
    expect(typeof attachSpy.mock.calls[0][2]).toBe('function');
  });

  describe('requestSpy', () => {
    it('is called with clientRequest', () => {
      const requestSpy = jest.fn();
      attachHttpRequestSpy(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(1);
      const spy = attachSpy.mock.calls[0][2];
      const callOriginal = jest.fn(() => 'client request object');
      spy(['http://example.tld'], callOriginal);
      expect(requestSpy).toHaveBeenCalledTimes(1);
      expect(requestSpy.mock.calls[0][0]).toBe('client request object');
    });

    it('is called with object options', () => {
      const requestSpy = jest.fn();
      attachHttpRequestSpy(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(1);
      const spy = attachSpy.mock.calls[0][2];
      const callOriginal = jest.fn(() => 'client request object');
      spy([{
        protocol: 'http',
        hostname: 'example.tld',
        port: 8080,
        method: 'GET',
        path: '/somewhere?over=rainbow#so-blue',
        auth: 'user:password',
      }], callOriginal);
      expect(requestSpy).toHaveBeenCalledTimes(1);
      expect(requestSpy.mock.calls[0][1]).toMatchSnapshot();
    });

    it('is called with sparse object options', () => {
      const requestSpy = jest.fn();
      attachHttpRequestSpy(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(1);
      const spy = attachSpy.mock.calls[0][2];
      const callOriginal = jest.fn(() => 'client request object');
      spy([{ method: 'GET' }], callOriginal);
      expect(requestSpy).toHaveBeenCalledTimes(1);
      expect(requestSpy.mock.calls[0][1]).toMatchSnapshot();
    });

    it('is called with parsed options', () => {
      const requestSpy = jest.fn();
      attachHttpRequestSpy(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(1);
      const spy = attachSpy.mock.calls[0][2];
      const callOriginal = jest.fn(() => 'client request object');
      spy(['http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue'], callOriginal);
      expect(requestSpy).toHaveBeenCalledTimes(1);
      expect(requestSpy.mock.calls[0][1]).toMatchSnapshot();
    });
  });

  describe('socketCloseSpy', () => {
    it('is called when the request socket closes', () => {
      onFinished.mockClear();
      const socketCloseSpy = jest.fn();
      attachHttpRequestSpy(jest.fn(), socketCloseSpy);
      expect(attachSpy).toHaveBeenCalledTimes(1);
      const spy = attachSpy.mock.calls[0][2];
      const clientRequest = {};
      const callOriginal = () => clientRequest;
      spy(['http://example.tld'], callOriginal);
      expect(onFinished).toHaveBeenCalledTimes(1);
      onFinished.mock.calls[0][1]();
      expect(socketCloseSpy).toHaveBeenCalledTimes(1);
      expect(socketCloseSpy.mock.calls[0][0]).toBe(clientRequest);
      expect(socketCloseSpy.mock.calls[0][1]).toMatchSnapshot();
    });
  });
});
