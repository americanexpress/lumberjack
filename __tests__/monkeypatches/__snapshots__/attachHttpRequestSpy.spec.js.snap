// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`attachHttpRequestSpy > requestSpy > is called with object options 1`] = `
Url {
  "auth": "user:password",
  "hash": "#so-blue",
  "host": "example.tld:8080",
  "hostname": "example.tld",
  "href": "http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue",
  "path": "/somewhere?over=rainbow",
  "pathname": "/somewhere",
  "port": "8080",
  "protocol": "http:",
  "query": "over=rainbow",
  "search": "?over=rainbow",
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy > requestSpy > is called with parsed options 1`] = `
Url {
  "auth": "user:password",
  "hash": "#so-blue",
  "host": "example.tld:8080",
  "hostname": "example.tld",
  "href": "http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue",
  "path": "/somewhere?over=rainbow",
  "pathname": "/somewhere",
  "port": "8080",
  "protocol": "http:",
  "query": "over=rainbow",
  "search": "?over=rainbow",
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy > requestSpy > is called with sparse object options 1`] = `
Url {
  "auth": null,
  "hash": null,
  "host": "localhost",
  "hostname": "localhost",
  "href": "http://localhost/",
  "path": "/",
  "pathname": "/",
  "port": null,
  "protocol": "http:",
  "query": null,
  "search": null,
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy > socketCloseSpy > is called when the request socket closes 1`] = `
Url {
  "auth": null,
  "hash": null,
  "host": "example.tld",
  "hostname": "example.tld",
  "href": "http://example.tld/",
  "path": "/",
  "pathname": "/",
  "port": null,
  "protocol": "http:",
  "query": null,
  "search": null,
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy > throws if requestSpy is not a function 1`] = `[TypeError: requestSpy must be a function (was "undefined")]`;

exports[`attachHttpRequestSpy > throws if socketCloseSpy is provided but is not a function 1`] = `[TypeError: socketCloseSpy must be function if provided (was "string")]`;

exports[`attachHttpRequestSpy requestSpy is called with object options 1`] = `
Url {
  "auth": "user:password",
  "hash": "#so-blue",
  "host": "example.tld:8080",
  "hostname": "example.tld",
  "href": "http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue",
  "path": "/somewhere?over=rainbow",
  "pathname": "/somewhere",
  "port": "8080",
  "protocol": "http:",
  "query": "over=rainbow",
  "search": "?over=rainbow",
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy requestSpy is called with parsed options 1`] = `
Url {
  "auth": "user:password",
  "hash": "#so-blue",
  "host": "example.tld:8080",
  "hostname": "example.tld",
  "href": "http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue",
  "path": "/somewhere?over=rainbow",
  "pathname": "/somewhere",
  "port": "8080",
  "protocol": "http:",
  "query": "over=rainbow",
  "search": "?over=rainbow",
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy requestSpy is called with sparse object options 1`] = `
Url {
  "auth": null,
  "hash": null,
  "host": "localhost",
  "hostname": "localhost",
  "href": "http://localhost/",
  "path": "/",
  "pathname": "/",
  "port": null,
  "protocol": "http:",
  "query": null,
  "search": null,
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy socketCloseSpy is called when the request socket closes 1`] = `
Url {
  "auth": null,
  "hash": null,
  "host": "example.tld",
  "hostname": "example.tld",
  "href": "http://example.tld/",
  "path": "/",
  "pathname": "/",
  "port": null,
  "protocol": "http:",
  "query": null,
  "search": null,
  "slashes": true,
}
`;

exports[`attachHttpRequestSpy throws if requestSpy is not a function 1`] = `"requestSpy must be a function (was "undefined")"`;

exports[`attachHttpRequestSpy throws if socketCloseSpy is provided but is not a function 1`] = `"socketCloseSpy must be function if provided (was "string")"`;
